### PR TITLE
🐛 Fix amp-consent sandbox to allow forms

### PR DESCRIPTION
**Pull request by JacopoBassan**
+ added allow-forms to amp-consent iframe sandbox

This change is to fix the impossibility to use any form, such as login forms, from any page opened from the amp-consent iframe where the CMP component can contain for example privacy policy links that can lead to the rest of the website were users can't login from.

### CODE REVIEW

**Comment by newmuis** [comment]
I think this will need to be assessed for security, as it's not immediately clear to me that this change is safe to make.  /cc @erwinmombay 

Anecdotally, I don't know that I would consider it a good user experience to navigate multiple pages _inside_ of the amp-consent iframe...

**Comment by gerstnr** [comment]
This case could be covered with https://github.com/ampproject/amphtml/pull/38250

**Comment by newmuis** [comment]
Could this be a good topic for an AMP design review?  Both #38250 and this PR loosen the security sandboxing of AMP and therefore should undergo security review and design review, per the [process for significant changes](https://github.com/ampproject/amphtml/blob/main/docs/contributing-code.md#process-for-significant-changes).

**Comment by 0x7f** [comment]
@newmuis @JacopoBassan incorporating the `allow-forms` attribute certainly enhances some functionalities, such as displaying a login form within the `amp-consent` layer. However, there are still other scenarios where issues may persist.

Consider a simple consent layer containing a "privacy policy" link. Using `target=_blank` opens the link in a new tab, allowing the user to read the policy in a full tab. Since the link opens from within an iframe with sandbox restrictions, the new tab inherits those restrictions, making the site appear dysfunctional to the user without an apparent cause.

To address this situation, the `allow-popups-to-escape-sandbox` sandbox attribute [was introduced](https://html.spec.whatwg.org/#attr-iframe-sandbox-allow-popups-to-escape-sandbox). This enables website owners to ensure that links opened from within the iframe function correctly.

As a result, I believe it is optimal to provide website owners with the flexibility to specify which sandbox attributes they want to enable. I'm eager to engage in a security review, as you suggested. Could you recommend a guide that would assist us in this pursuit?

**Comment by JacopoBassan** [comment]
> Consider a simple consent layer containing a "privacy policy" link. Using `target=_blank` opens the link in a new tab, allowing the user to read the policy in a full tab. Since the link opens from within an iframe with sandbox restrictions, the new tab inherits those restrictions, making the site appear dysfunctional to the user without an apparent cause.
> 
> To address this situation, the `allow-popups-to-escape-sandbox` sandbox attribute [was introduced](https://html.spec.whatwg.org/#attr-iframe-sandbox-allow-popups-to-escape-sandbox). This enables website owners to ensure that links opened from within the iframe function correctly.
> 
> As a result, I believe it is optimal to provide website owners with the flexibility to specify which sandbox attributes they want to enable. I'm eager to engage in a security review, as you suggested. Could you recommend a guide that would assist us in this pursuit?

I actually didn't know about the `allow-popups-to-escape-sandbox` attribute, that would easily and more rapidly solve the issues on our side with less security concern on the AMP part, I definitely route for that. I do also agree that the flexibility to let website owners define those attributes like already happens in the amp-iframe config would be better.

**Comment by JacopoBassan** [comment]
I updated my code, I think the `allow-popups-to-escape-sandbox` would be the best route, and as I said should lead to less security-wise concerns. I also updated the test files.

@newmuis please let me know if you think this could go through, thanks!

**Comment by newmuis** [comment]
This is really helpful, thanks.

On the surface I would hope this can resolve things, but I'm actually curious why browsers have this permission and apply sandboxing to popups in the first place.  Does this imply that popups have permissions on the host page (maybe via `window.opener` or similar)?  How can we ensure that enabling this permission doesn't actually give popups escalated privileges on AMP pages?

Given that it's still an explicit opt-in from publishers, this might be ok, but we should still understand the implications of this before granting that privilege.

**Comment by JacopoBassan** [comment]
@newmuis thanks Jon for the reply, I agree, and I've found some interesting infos looking around.

About` allow-popups-to-escape-sandbox` ([source](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/sandbox))
> Allows a sandboxed document to open new windows without forcing the sandboxing flags upon them. This will allow, for example, a third-party advertisement to be safely sandboxed without forcing the same restrictions upon the page the ad links to.

And about window.opener ([source](https://developer.mozilla.org/en-US/docs/Web/API/Window/opener))
> If the opener is not on the same origin as the current page, functionality of the opener object is limited. For example, variables and functions on the window object are not accessible. However, navigation of the opener window is possible, which means that the opened page can open a URL in the original tab or window. In some cases, this makes phishing attacks possible, where a trusted page that is opened in the original window is replaced by a phishing page by the newly opened page.
> 
> In the following cases, the browser does not populate window.opener, but leaves it [null](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/null):
> 
> - The opener can be omitted by specifying [rel](/en-US/docs/Web/HTML/Element/a#rel)=noopener on a link, or passing noopener in the [windowFeatures](https://developer.mozilla.org/en-US/docs/Web/API/Window/open) parameter.
> - Windows opened because of links with a [target](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#target) of _blank don't get an opener, unless explicitly requested with [rel](/en-US/docs/Web/HTML/Element/a#rel)=opener.
> - Having a [Cross-Origin-Opener-Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Opener-Policy) header with a value of same-origin prevents setting opener. Since the new window is loaded in a different browsing context, it won't have a reference to the opening window.

As probably we all expected is mainly sandboxed to safely include ads and in general external content we could not have control over, since yes the window.opener could be something to create issues where we don't know the source.

In our case though the new window doesn't get the window.opener populated since we are using target _blank to open in a new tab, and at best if it was we would have access to the amp-consent iframe not the AMP page, but the amp-consent iframe only loads the promptUISrc which is under our control so we know what to expect from that page anyway.

**Comment by erwinmombay** [comment]
@JacopoBassan im writing a quick google doc just to make sure i understand the problem and solution, would you mind if i send it your way

**Comment by JacopoBassan** [comment]
@erwinmombay sure, anything that can help make this through

**Comment by erwinmombay** [comment]
@JacopoBassan is there any deadline date that CMP's need to meet? I'm trying to get a review as quickly as possible internally and want to convey any timelines to the security engineer to get this going quicker

**Comment by JacopoBassan** [comment]
@erwinmombay thank you so much, there's not a defined deadline but it's a "the sooner the better" situation where this is both a current problem in production and also is blocking a new development since quite some time.

**Comment by JacopoBassan** [comment]
@erwinmombay any news on this? It's been 2 weeks and we are in a hurry to have this solved, thanks.

**Comment by erwinmombay** [comment]
@JacopoBassan sorry for the wait and thank you for being patient. security just approved these changes so I am merging this

**Review by erwinmombay — APPROVED** [review]

**Comment by JacopoBassan** [comment]
@erwinmombay thanks for the great news, usually after the approval when we should expect to see this change in production?

**Comment by erwinmombay** [comment]
@JacopoBassan it looks like the PR might be stale. do you mind rebasing it with the latest main and re-pushing the code? thanks. ill check on this again tomorrow morning

**Comment by JacopoBassan** [comment]
@erwinmombay done, thanks

**Comment by erwinmombay** [comment]
@JacopoBassan , looks like theres an error in the Circle CI run. would you mind rebasing again.

File: extensions/amp-consent/0.1/test/test-consent-ui.js
Comment (erwinmombay) [inline extensions/amp-consent/0.1/test/test-consent-ui.js:149]: would you mind adding documentation around this

Comment (erwinmombay) [reply extensions/amp-consent/0.1/test/test-consent-ui.js:149]: we just merged https://github.com/ampproject/amphtml/pull/38250/files, there should be a new "sandbox" property to add these new values. would you mind adding it on to there and then adding docs. 

If not, i can fork this PR to add this request

Comment (erwinmombay) [reply extensions/amp-consent/0.1/test/test-consent-ui.js:149]: to @JacopoBassan

Comment (JacopoBassan) [reply extensions/amp-consent/0.1/test/test-consent-ui.js:149]: @erwinmombay if you could do that it would be great, I’m on vacation this week so I’m not able to do that, thanks!

Comment (erwinmombay) [reply extensions/amp-consent/0.1/test/test-consent-ui.js:149]: looks like the other PR i just merged already covers the `allow-popups-to-escape-sandbox` case https://github.com/ampproject/amphtml/pull/38250/files so this should be live in 2 weeks. do verify it though if things look good to you. thank you!

Comment (JacopoBassan) [reply extensions/amp-consent/0.1/test/test-consent-ui.js:149]: @erwinmombay looks good to me thanks, let's wait for the go live

**Comment by JacopoBassan** [comment]
Closed because #38250 cover this and has been merged.

File: extensions/amp-consent/0.1/test/test-consent-ui.js
Comment (JacopoBassan) [reply extensions/amp-consent/0.1/test/test-consent-ui.js:149]: @erwinmombay hi, any updates? I'm still not seeing the update in production, have you had any delays? We need to have this asap. Thanks.

**Pull request closed without merging**
